R_notebooks/04_Piping.Rmd

```diff
@@ -0,0 +1,132 @@
+output: html_notebook
+---
+
+## Outline
+
+- The functional philosophy of Tidyverse
+- Piping operator
+
+
+## The Functional philosophy of Tidyverse
+
+The Tidyverse packages are built around a philosophy of having functions that
+work well together. Let's look at some of the functions we've seen so far:
+
+| Package | Function    | Data input type | Output type       |
+|---------|-------------|-----------------|-------------------|
+| `tidyr` | `read_csv`  | string          | dataframe         |
+| `dplyr` | `filter`    | dataframe       | dataframe         |
+| `dplyr` | `select`    | dataframe       | dataframe         |
+| `dplyr` | `mutate`    | dataframe       | dataframe         |
+| `dplyr` | `relocate`  | dataframe       | dataframe         |
+| `dplyr` | `arrange`   | dataframe       | dataframe         |
+| `dplyr` | `summarise` | dataframe       | dataframe         |
+| `dplyr` | `group_by`  | dataframe       | grouped dataframe |
+
+Notice how consistent these are in accepting a dataframe as input and returning
+a dataframe (or something like a dataframe). It is this consistency that makes
+it easy to compose functions. 
+
+In addition, it's important to note that none of the functions we've seen so
+far modify the input dataframe 'in place', they always return a new dataframe.
+(This is even the case for the `mutate` function, despite the name!) This
+encourages a 'pipeline' approach, where we successively take the output
+of one function and feed it in as the input to another function.
+
+
+## Piping operator
+
+Since version 4.1, R has had a pipe operator `|>` which facilitates chaining
+function calls.
+
+**Note:** Before a pipe operator was introduced into the base R language, people
+used another pipe operator `%>%` from the `magrittr` package. It is still very
+common to see people using the `magrittr` pipe and it behaves in a similar way
+to base R's `|>` (though there are some minor differences). You can use
+whichever you like, though for this course we'll use the base R pipe.
+
+The pipe operator takes the value in the left hand side and 'feeds it into' the
+_first_ argument of a function call on the right hand side. So, using
+`dplyr::select` as an example, the following two expressions are equivalent ways
+to select two columns `colA` and `colB` from a dataframe `data`:
+
+```
+dplyr::select(data, colA, colB)     data |> dplyr::select(colA, colB)
+```
+
+Notice how it looks like the dataframe argument to `dplyr::select` is missing
+in the second expression. The pipe takes care of putting `data` into the first
+argument slot, and the remaining slots are available for all other arguments
+(in this case, the column names  `colA` and `colB`).
+
+We can now chain multiple pipe calls together, like so:
+
+```
+data |> dplyr::select(colA, colB) |> dplyr::filter(colA > 10)
+```
+
+The result of the whole of the expression to the right of the assignment `<-`
+gets stored in the `transformed_data`. In words, the above
```

Comment:
transformed_data hasn't been assigned in line 67 and the expression is therefore not to the right of the assignment. Consider rewording the description or adding the assignment into line 67 of code.